File locking after .close() on Windows.

When running unit tests on Windows, StoreWALTest fails. It seems to be related to this : http://stackoverflow.com/questions/4179145/release-java-file-lock-in-windows.

Maybe it's not possible to solve properly, but maybe implement a windows workaround to fix this issue ...

I'm using a French OS, so here's the IOException message translated in English : The requested operation cannot be performed on a file with a user-mapped section open.

```
Tests run: 47, Failures: 1, Errors: 2, Skipped: 0, Time elapsed: 3.881 sec <<< FAILURE! - in org.mapdb.StoreWALTest
delete_files_after_close2(org.mapdb.StoreWALTest)  Time elapsed: 0.095 sec  <<< FAILURE!
java.lang.AssertionError: null
    at org.junit.Assert.fail(Assert.java:86)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at org.junit.Assert.assertFalse(Assert.java:64)
    at org.junit.Assert.assertFalse(Assert.java:74)
    at org.mapdb.StoreWALTest.delete_files_after_close2(StoreWALTest.java:52)

large_record_larger(org.mapdb.StoreWALTest)  Time elapsed: 2.779 sec  <<< ERROR!
java.io.IOError: java.io.IOException: L’opération demandée n’a pu s’accomplir sur un fichier ayant une section mappée utilisateur ouverte
    at org.mapdb.Volume$MappedFileVol.truncate(Volume.java:537)
    at org.mapdb.StoreWAL.logReset(StoreWAL.java:124)
    at org.mapdb.StoreWAL.replayLogFile(StoreWAL.java:863)
    at org.mapdb.StoreWAL.commit(StoreWAL.java:621)
    at org.mapdb.EngineTest.large_record_larger(EngineTest.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
    at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
Caused by: java.io.IOException: L’opération demandée n’a pu s’accomplir sur un fichier ayant une section mappée utilisateur ouverte
    at sun.nio.ch.FileDispatcherImpl.truncate0(Native Method)
    at sun.nio.ch.FileDispatcherImpl.truncate(FileDispatcherImpl.java:92)
    at sun.nio.ch.FileChannelImpl.truncate(FileChannelImpl.java:346)
    at org.mapdb.Volume$MappedFileVol.truncate(Volume.java:535)
    at org.mapdb.StoreWAL.logReset(StoreWAL.java:124)
    at org.mapdb.StoreWAL.replayLogFile(StoreWAL.java:863)
    at org.mapdb.StoreWAL.commit(StoreWAL.java:621)
    at org.mapdb.EngineTest.large_record_larger(EngineTest.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
    at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)

put_reopen_get_large(org.mapdb.StoreWALTest)  Time elapsed: 0.033 sec  <<< ERROR!
java.io.IOError: java.io.IOException: L’opération demandée n’a pu s’accomplir sur un fichier ayant une section mappée utilisateur ouverte
    at org.mapdb.Volume$MappedFileVol.truncate(Volume.java:537)
    at org.mapdb.StoreWAL.logReset(StoreWAL.java:124)
    at org.mapdb.StoreWAL.replayLogFile(StoreWAL.java:863)
    at org.mapdb.StoreWAL.commit(StoreWAL.java:621)
    at org.mapdb.EngineTest.put_reopen_get_large(EngineTest.java:64)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
    at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
Caused by: java.io.IOException: L’opération demandée n’a pu s’accomplir sur un fichier ayant une section mappée utilisateur ouverte
    at sun.nio.ch.FileDispatcherImpl.truncate0(Native Method)
    at sun.nio.ch.FileDispatcherImpl.truncate(FileDispatcherImpl.java:92)
    at sun.nio.ch.FileChannelImpl.truncate(FileChannelImpl.java:346)
    at org.mapdb.Volume$MappedFileVol.truncate(Volume.java:535)
    at org.mapdb.StoreWAL.logReset(StoreWAL.java:124)
    at org.mapdb.StoreWAL.replayLogFile(StoreWAL.java:863)
    at org.mapdb.StoreWAL.commit(StoreWAL.java:621)
    at org.mapdb.EngineTest.put_reopen_get_large(EngineTest.java:64)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
    at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
```

--- a/src/main/java/org/mapdb/StoreWAL.java
+++ b/src/main/java/org/mapdb/StoreWAL.java
@@ -68,7 +68,6 @@ public class StoreWAL extends StoreDirect {
             }
             replayPending = false;
             checkHeaders();
-            log = volFac.createTransLogVolume();
             if(!readOnly)
                 logReset();
             allGood = true;

--- a/src/main/java/org/mapdb/Volume.java
+++ b/src/main/java/org/mapdb/Volume.java
@@ -402,6 +402,9 @@ public abstract class Volume {
             }
         }
 
+        // Workaround for https://github.com/jankotek/MapDB/issues/326
+        // File locking after .close() on Windows.
+        private static boolean windowsWorkaround = System.getProperty("os.name").toLowerCase().startsWith("win");
 
     }
 
@@ -531,10 +534,23 @@ public abstract class Volume {
                     old[i] = null;
                 }
 
+                if (ByteBufferVol.windowsWorkaround) {
+                    for(int i=0;i<maxSize;i++){
+                        unmap((MappedByteBuffer) old[i]);
+                        old[i] = null;
+                    }
+                }
+
                 try {
                     fileChannel.truncate(1L * chunkSize *maxSize);
                 } catch (IOException e) {
                     throw new IOError(e);
+                }
+
+                if (ByteBufferVol.windowsWorkaround) {
+                    for(int pos=0;pos<maxSize;pos++) {
+                        chunks[pos]=makeNewBuffer(1L*chunkSize*pos);
+                    }
                 }
 
             }finally {
@@ -626,6 +642,7 @@ public abstract class Volume {
 
         protected final File file;
         protected final int chunkSize;
+        protected RandomAccessFile raf;
         protected FileChannel channel;
         protected final boolean readOnly;
         protected final long sizeLimit;
@@ -643,10 +660,12 @@ public abstract class Volume {
             try {
                 checkFolder(file,readOnly);
                 if(readOnly && !file.exists()){
+                    raf = null;
                     channel = null;
                     size = 0;
                 }else {
-                    channel = new RandomAccessFile(file, readOnly ? "r" : "rw").getChannel();
+                    raf = new RandomAccessFile(file, readOnly ? "r" : "rw");
+                    channel = raf.getChannel();
                     size = channel.size();
                 }
             } catch (IOException e) {
@@ -859,6 +878,9 @@ public abstract class Volume {
                 if(channel!=null)
                     channel.close();
                 channel = null;
+                if (raf != null)
+                    raf.close();
+                raf = null;
             }catch(IOException e){
                 throw new IOError(e);
             }

